### PR TITLE
ACLs: Add migration flow on startup

### LINKED ISSUE #8089: ACLs: Check migration status on startup
Verify if SuperAdmin user is setup on startup.
If not, check if migration flow is needed.
If multiple users / credentials - refer to superuser command

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #8089

## Change Description

Add basic auth migration flow on startup

## Testing
Added unit tests for migration code

Tested manually all combinations for migration + manual import on lakefs run

### CODE REVIEW

File: pkg/auth/basic_service_test.go [hunk 1]
```diff
@@ -129,29 +129,9 @@ func TestBasicAuthService_CredentialsImport(t *testing.T) {
...
-	creds.AccessKeyId = "A" + secretAccessKey
-	encryptedKey, err = model.EncryptSecret(s.SecretStore(), "BadSecret")
-	require.NoError(t, err)
-
-	creds.SecretAccessKeyEncryptedBytes = encryptedKey
-	credsData, err = proto.Marshal(creds)
-	require.NoError(t, store.Set(ctx, []byte(model.PartitionKey), model.CredentialPath(user.Username, creds.AccessKeyId), credsData))
+	user := createOldUser(t, ctx, store, "user-old")
+	createOldCreds(t, ctx, store, s, user.Username, accessKeyID, secretAccessKey)
+	createOldCreds(t, ctx, store, s, user.Username, "A"+accessKeyID, "BadSecret")
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/auth/basic_service_test.go:134]: It's just to ensure we are migrating the right credentials

File: cmd/lakefs/cmd/run.go [hunk 2]
```diff
@@ -83,6 +83,30 @@ func NewAuthService(ctx context.Context, cfg *config.Config, logger logging.Logg
...
+		// Check if migration needed
+		initialized, err := metadataManager.IsInitialized(ctx)
+		if err != nil {
+			logger.WithError(err).Fatal("failed to get lakeFS init status")
+		}
+		if initialized {
+			username, err := apiService.Migrate(ctx)
+			switch {
+			case errors.Is(err, auth.ErrMigrationNotPossible):
+				logger.WithError(err).Fatal("\n**********************************\n" +
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply cmd/lakefs/cmd/run.go:95]: 1. We have it only in lakectl
2. Done - I'm just worried the message is going to be hard to detect
3. Done

File: pkg/auth/basic_service.go [hunk 3]
```diff
@@ -46,6 +46,62 @@ func NewBasicAuthService(store kv.Store, secretStore crypt.SecretStore, cacheCon
 	return res
 }
 
+// Migrate - checks lakeFS server status for migration to basic auth
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/auth/basic_service.go:49]: Refactored

File: pkg/auth/basic_service.go [hunk 4]
```diff
@@ -46,6 +46,62 @@ func NewBasicAuthService(store kv.Store, secretStore crypt.SecretStore, cacheCon
...
+			username, err := s.CreateUser(ctx, user)
+			return username, err
+		default: // If more than one user defined in system - user must run migration manually
+			return "", fmt.Errorf("too many users: %w", ErrMigrationNotPossible)
+		}
+	}
+	return "", err
+}
+
+func (s *BasicAuthService) listUserForMigration(ctx context.Context) ([]*model.User, error) {
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/auth/basic_service.go:77]: We don't, hence:
`for len(entries) <= MaxUsers && it.Next() {`

File: cmd/lakefs/cmd/run.go [hunk 5]
```diff
@@ -83,6 +83,30 @@ func NewAuthService(ctx context.Context, cfg *config.Config, logger logging.Logg
...
+					"cannot migrate existing user to basic auth mode!\n" +
+					"Please run `lakefs superuser -h` and follow the instructions on how to migrate an existing user\n" +
+					"**********************************\n")
+			case err == nil:
+				if username != "" { // Print only in case of actual migration
+					logger.Warning("\n**********************************\n" +
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply cmd/lakefs/cmd/run.go:101]: Logs are outputted by default to Stderr, I'll change it to Info

**Review by arielshaqed — APPROVED** [review]
Great, thanks!

> I've considered this edge case and decided this is not something we should enforce. If the user ultimately decides he wants to user a non admin user as his "superAdmin" user - I don't want to block that. We can request @treeverse/product to provide their input

I think I'm worried about the opposite.  You set up lakeFS for me with me as the single user, who is _not_ an admin.  I trick you into upgrading lakeFS.  Now I am the single user on the upgraded lakeFS, and I _am_ an admin.

Please communicate this to @treeverse/product.  It is of course unavoidable with "auto-migrate on upgrade", but they should be aware.

**Comment by N-o-Z** [comment]
> Great, thanks!
> 
> > I've considered this edge case and decided this is not something we should enforce. If the user ultimately decides he wants to user a non admin user as his "superAdmin" user - I don't want to block that. We can request @treeverse/product to provide their input
> 
> I think I'm worried about the opposite. You set up lakeFS for me with me as the single user, who is _not_ an admin. I trick you into upgrading lakeFS. Now I am the single user on the upgraded lakeFS, and I _am_ an admin.
> 
> Please communicate this to @treeverse/product. It is of course unavoidable with "auto-migrate on upgrade", but they should be aware.

The scenario of having a single user in the system which is not an Admin is highly unlikely. Also running the lakefs superuser command require access to the deployment environment so I'm not that concerned from the scenario you suggested.
Regarding @treeverse/product I think we tagged them here enough times so they could have provided their input if they felt it's needed

**Pull request merged**